vscode/src/completions/context/context-data-logging.ts

```diff
@@ -0,0 +1,144 @@
+import {
+    type AutocompleteContextSnippet,
+    isDotComAuthed,
+    subscriptionDisposable,
+} from '@sourcegraph/cody-shared'
+import type * as vscode from 'vscode'
+import { logDebug } from '../../log'
+import { GitHubDotComRepoMetadata } from '../../repository/repo-metadata-from-git-api'
+import { completionProviderConfig } from '../completion-provider-config'
+import type { ContextRetriever } from '../types'
+import type { RetrievedContextResults } from './completions-context-ranker'
+import { JaccardSimilarityRetriever } from './retrievers/jaccard-similarity/jaccard-similarity-retriever'
+import { DiagnosticsRetriever } from './retrievers/recent-user-actions/diagnostics-retriever'
+import { RecentCopyRetriever } from './retrievers/recent-user-actions/recent-copy'
+import { RecentEditsRetriever } from './retrievers/recent-user-actions/recent-edits-retriever'
+import { RecentViewPortRetriever } from './retrievers/recent-user-actions/recent-view-port'
+import { RetrieverIdentifier } from './utils'
+
+interface RetrieverConfig {
+    identifier: RetrieverIdentifier
+    maxSnippets: number
+}
+
+export class ContextRetrieverDataCollection implements vscode.Disposable {
+    public dataCollectionRetrievers: ContextRetriever[] = []
+    private disposables: vscode.Disposable[] = []
+    private static readonly MAX_PAYLOAD_SIZE_BYTES = 1024 * 1024 // 1 MB
+    private dataCollectionFlagState = false
+
+    private readonly retrieverConfigs: RetrieverConfig[] = [
+        { identifier: RetrieverIdentifier.RecentCopyRetriever, maxSnippets: 1 },
+        { identifier: RetrieverIdentifier.RecentEditsRetriever, maxSnippets: 15 },
+        { identifier: RetrieverIdentifier.DiagnosticsRetriever, maxSnippets: 15 },
+        { identifier: RetrieverIdentifier.RecentViewPortRetriever, maxSnippets: 10 },
+        { identifier: RetrieverIdentifier.JaccardSimilarityRetriever, maxSnippets: 15 },
+    ]
+
+    constructor() {
+        this.disposables.push(
+            subscriptionDisposable(
+                completionProviderConfig.completionDataCollectionFlag.subscribe(
+                    this.manageDataCollectionRetrievers.bind(this)
+                )
+            )
+        )
+    }
+
+    private manageDataCollectionRetrievers(dataCollectionEnabled: boolean): void {
+        if (this.dataCollectionFlagState === dataCollectionEnabled) {
+            return
+        }
+
+        this.dataCollectionFlagState = dataCollectionEnabled
+        this.disposeDataCollectionRetrievers()
+
+        if (dataCollectionEnabled) {
+            this.dataCollectionRetrievers = this.retrieverConfigs
+                .map(this.createRetriever)
+                .filter((retriever): retriever is ContextRetriever => retriever !== undefined)
```

Comment:
We could use `isDefined` from `@sourcegraph/cody-shared` here:

```suggestion
                .filter(isDefined)
```

Reply:
done